File: qc-application/src/main/java/com/eager/questioncloud/application/business/payment/point/implement/ChargePointPaymentPostProcessor.java
package com.eager.questioncloud.application.business.payment.point.implement;

import com.eager.questioncloud.application.business.payment.point.event.ChargePointPaymentEvent;
import com.eager.questioncloud.application.message.FailChargePointPaymentMessage;
import com.eager.questioncloud.application.message.MessageSender;
import com.eager.questioncloud.application.message.MessageType;
import com.eager.questioncloud.core.domain.point.implement.UserPointManager;
import com.eager.questioncloud.core.exception.CoreException;
import com.eager.questioncloud.core.exception.Error;
import lombok.RequiredArgsConstructor;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class ChargePointPaymentPostProcessor {
    private final UserPointManager userPointManager;
    private final MessageSender messageSender;

    @EventListener
    public void chargeUserPoint(ChargePointPaymentEvent event) {
        try {
            userPointManager.chargePoint(event.getUserId(), event.getChargePointType().getAmount());
        } catch (Exception e) {
            messageSender.sendMessage(MessageType.FAIL_CHARGE_POINT, FailChargePointPaymentMessage.create(event.getPaymentId()));
            throw new CoreException(Error.PAYMENT_ERROR);
        }
    }
}

File: qc-application/src/main/java/com/eager/questioncloud/application/business/payment/point/implement/ChargePointPaymentPostProcessor.kt
package com.eager.questioncloud.application.business.payment.point.implement

import com.eager.questioncloud.application.business.payment.point.event.ChargePointPaymentEvent
import com.eager.questioncloud.core.domain.point.implement.UserPointManager
import io.awspring.cloud.sqs.annotation.SqsListener
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.stereotype.Component

@Component
class ChargePointPaymentPostProcessor(
    private val userPointManager: UserPointManager,
) {
    @SqsListener("charge-point.fifo")
    fun chargeUserPoint(@Payload event: ChargePointPaymentEvent) {
        userPointManager.chargePoint(event.userId, event.chargePointType.amount)
    }
}
